cognite/client/data_classes/data_modeling/containers.py

```diff
@@ -73,110 +95,83 @@ def as_property_ref(self, property: str) -> tuple[str, str, str]:
         return self.as_id().as_property_ref(property)
 
 
+@dataclass(frozen=True)
 class ContainerApply(ContainerCore):
     """Represent the physical storage of data. This is the write format of the container
 
     Args:
         space (str): The workspace for the container, a unique identifier for the space.
         external_id (str): Combined with the space is the unique identifier of the container.
-        properties (dict[str, ContainerProperty]): We index the property by a local unique identifier.
+        properties (Mapping[str, ContainerPropertyApply]): We index the property by a local unique identifier.
         description (str | None): Textual description of the container
         name (str | None): Human readable name for the container.
         used_for (Literal['node', 'edge', 'all'] | None): Should this operation apply to nodes, edges or both.
-        constraints (dict[str, Constraint] | None): Set of constraints to apply to the container
-        indexes (dict[str, Index] | None): Set of indexes to apply to the container.
+        constraints (Mapping[str, ConstraintApply] | None): Set of constraints to apply to the container
+        indexes (Mapping[str, IndexApply] | None): Set of indexes to apply to the container.
     """
 
-    def __init__(
-        self,
-        space: str,
-        external_id: str,
-        properties: dict[str, ContainerProperty],
-        description: str | None = None,
-        name: str | None = None,
-        used_for: Literal["node", "edge", "all"] | None = None,
-        constraints: dict[str, Constraint] | None = None,
-        indexes: dict[str, Index] | None = None,
-    ) -> None:
-        validate_data_modeling_identifier(space, external_id)
-        super().__init__(space, external_id, properties, description, name, constraints, indexes)
-        self.used_for = used_for
+    properties: Mapping[str, ContainerPropertyApply]
+    constraints: Mapping[str, ConstraintApply] = field(default_factory=dict)
+    indexes: Mapping[str, IndexApply] = field(default_factory=dict)
+    used_for: Literal["node", "edge", "all"] | None = None
+
+    def __post_init__(self) -> None:
+        validate_data_modeling_identifier(self.space, self.external_id)
 
     @classmethod
     def _load(cls, resource: dict, cognite_client: AsyncCogniteClient | None = None) -> ContainerApply:
-        return ContainerApply(
-            space=resource["space"],
-            external_id=resource["externalId"],
-            properties={k: ContainerProperty.load(v) for k, v in resource["properties"].items()},
-            description=resource.get("description"),
-            name=resource.get("name"),
+        return cls._load_base(
+            resource,
+            container_class=cls,
+            property_loader=ContainerPropertyApply._load,
+            constraint_loader=ConstraintApply._load,
+            index_loader=IndexApply._load,
```

Comment:
We seem to be always calling `._load` on the passed callables - then I think I would prefer to pass the classes instead.

Reply:
The problem was with the `Constraint` and `Index` classes. If we switch the argument to `constraint_class: type[ConstraintCore]` and pass `constraint_class=Constraint` or `constraint_class=ConstraintApply`, mypy will complain with
```
cognite/client/data_classes/data_modeling/containers.py:127: error: Only concrete class can be given where "type[ConstraintCore]" is expected  [type-abstract]
```
So I switched to using loader functions for these arguments.

We can still pass classes for the property. Should we?